sensor-vernier/src/main/java/org/concord/sensor/vernier/d2pio/D2PIOSensorDevice.java

```diff
@@ -0,0 +1,293 @@
+			return null;
+		}
+		ExperimentConfigImpl expConfig = new ExperimentConfigImpl();
+
+		expConfig.setDeviceName(currentGoDirectDevice.getDeviceLabel());
+
+		expConfig.setExactPeriod(true);
+		expConfig.setPeriod((float)currentGoDirectDevice.getMeasurementPeriod());
+		expConfig.setDataReadPeriod(expConfig.getPeriod());
+
+		// TODO we should set the period range since it is known by the device
+		//   expConfig.getPeriodRange();
+
+		int numChannels = 0;
+		int channelMask = currentGoDirectDevice.getMeasurementChannelAvailabilityMask();
+
+		// Use channel mask to count number of channels
+		for (int ch = 0; ch < MAX_D2PIO_CHANNELS; ch++) {
+			if (((1 << ch) & channelMask) != 0) {
+				numChannels++;
+			}
+		}
+
+		SensorConfig [] sensorConfigs = new SensorConfig[numChannels];
+		expConfig.setSensorConfigs(sensorConfigs);
+		int chIndex = 0;
+		for (int ch = 0; ch < MAX_D2PIO_CHANNELS; ch++) {
+			if (((1 << ch) & channelMask) != 0) {
+				// TODO: what should the channel type be?
+				// Is it always CHANNEL_TYPE_ANALOG or is it ever CHANNEL_TYPE_DIGITAL?
+				int channelType = VernierSensor.CHANNEL_TYPE_ANALOG;
+				VernierSensor sensor = new VernierSensor(this, devService, chIndex, channelType);
+
+				// FIXME if the sensor isn't a known sensor then sensorConfig should return
+				// an unknown type of sensor.  It isn't clear if this should return a valid experiment config or not.
+				int sensorid = currentGoDirectDevice.getAttachedSensorId(ch);
+
+				sensor.setupSensor(currentGoDirectDevice.getAttachedSensorId(ch), null);
+				sensorConfigs[chIndex] = sensor;
+				chIndex++;
+			}
+		}
+
+		expConfig.setValid(true);
+
+		return expConfig;
+	}
+
+	protected D2PIOSensor openGoDirectDevice() {
+		if(d2pio == null){
+			return null;
+		}
+    boolean isDeviceAttached = d2pio.isSensorAttached(); //obsolete?
+
+		D2PIOSensor gSensor = d2pio.getFirstSensor();
+
+		if(gSensor == null){
+			// Set the error message here because returning null here
+			// ought to trigger an error printout, however if isAttached was called
+			// first then this shouldn't happen because that ought to return false
+			// in this case
+			errorMessage = "Cannot find an attached Go Direct device";
+			return null;
+		}
+
+		// This will lock the device to this thread
+		// In the sensor-native code we then unlock it.  It isn't clear why,
+		// perhaps so any thread can access it.  A more safe approach would be to use
+		// the a single thread delegator to force all access on one thread.
+		// or to synchronize the access to the gSensor and have it lock it and unlock it
+		gSensor.open();
+		return gSensor;
+	}
+
+	public String getErrorMessage(int error) {
+		if(errorMessage == null){
+			return "Unknown Error";
+		}
+
+		return errorMessage;
+	}
+
+	public String getVendorName() {
+		return "Vernier";
+	}
+
+	public String getDeviceName() {
+		if(currentGoDirectDevice != null){
+			return currentGoDirectDevice.getDeviceLabel();
+		}
+
+		return "GoDirect";
```

Comment:
What's the relationship between the device name and the device label?

Reply:
Device label appears to be a shortened version of the device name (2-3 char).  I'm voting "GDX" for device label and "Go Direct" with a space for device name.  The space is how Vernier does it on their promotional materials:
https://www.vernier.com/products/sensors/go-direct-sensors/